### ISSUE TITLE
PostgreSQL using kinit with Kerberos returns error "Unable to obtain password from user"

### ISSUE THREAD

**Issue by igor2x** [dbeaver/dbeaver#20066]
### Description

I would like to establish connection to PostgreSQL using Kerberos and kinit option, but in DBeaver logs I get error: "Unable to obtain password from user" which is little bit strange error, because if using kinit, that means using existing TGT ticket instead of a password.

### DBeaver Version

Lite Edition Version 23.0.0.202303201558

### Operating System

Ubuntu 23.04

### Database and driver

1. Database name and version: PostgreSQL 15.3
2. Driver name and version: default driver from DBeaver

### Steps to reproduce

1. Menu **Database | New Database Connection**.
2. Select **PostgreSQL** and **Next** button.
3. In **Driver properties** add new property.
![image](https://github.com/dbeaver/dbeaver/assets/10895030/dbddfc57-e136-4979-aca2-30b886ab2bfb)
5. Click on **Main** tab and fill out the dialog.
![image](https://github.com/dbeaver/dbeaver/assets/10895030/29f25136-ab25-4b01-9bac-15e782567a37)
6. Click on **Test Connection** and connection is successful.
7. Close down the dialogs.
8. Menu **SQL Editor | New SQL Script**
9. Paste the SQL: 
select pid, gss_authenticated, encrypted, principal from pg_stat_gssapi where pid = pg_backend_pid();
and execute the SQL.
![image](https://github.com/dbeaver/dbeaver/assets/10895030/c1199e49-7522-4fb1-91a0-3bcad281ff43)
9. In above picture second and third columns are both "true", so Kerberos connection is properly established.
So far no issue.

But I don't want to use _password_, because whole purpose of Kerbeors is to use tickets instead of passwords.
10. Open terminal and create Kerberos TGT ticket with `kinit` command and connect to PostgreSQL **without a password.**
`psql "postgresql:///user01?host=server.example.com&port=5432&dbname=postgres&gssencmode=require" -c "select pid, gss_authenticated, encrypted, principal from pg_stat_gssapi where pid = pg_backend_pid();"`

Connection was established successfully without password (using TGT ticket). Result of select statement is returned.

![image](https://github.com/dbeaver/dbeaver/assets/10895030/6781ec9e-19cf-4cd3-b8c8-df2dd007fc1d)

Now do the same in DBeaver. I have checked the documentation: https://github.com/dbeaver/dbeaver/wiki/Kerberos-authentication which is little bit outdated, but I can still follow it.
11. Edit existing connection.
12. Remove password from **Password** field.
13. Check the **Use kinit** checkbox. **Password** field gets greyed out, what is expected, because TGT ticket should be used instead of password.
14. Check the **Debug Kerberos connection** checkbox to get some useful info in DBeaver debug log.
15. Click on **Test connection** button but now problem appears, I get error message: "_An error occurred while setting up the GSS Encoded connection._"
16. From the DBeaver logs related to Kerbeos

> null credentials from Ticket Cache
> [Krb5LoginModule] authentication failed 
> Unable to obtain password from user
> 

Message "Unable to obtain password from user" is little bit strange to me. Whole purpose of Kerberos is to not use passwords, but tickets instead.

DBeaver debug log: [dbeaver-debug.log](https://github.com/dbeaver/dbeaver/files/11551971/dbeaver-debug.log)

### Additional context

"PostgreSQL client" and "PostgreSQL server" are running on separate Linux computers.

**Comment by LonwoLonwo** [comment]
Hello @igor2x 

Could you please try our Early Acces version? https://dbeaver.com/files/ea/lite/
We had some changes inside the Kerberos Authentication code in DBeaver.

**Comment by igor2x** [comment]
@LonwoLonwo I have installed Version 23.0.4.202305011233 as requested.

Now problem appears even sooner:
At step 5 connection is unsuccessful.
![image](https://github.com/dbeaver/dbeaver/assets/10895030/859a461e-bc92-41f8-a65b-2fb388bec236)

Steps 6 - 9 I can't perform now because above error.

Then from steps 11 - 15 and in step 15 returns the same error message as reported and I see the same in step 16 in debug logs.

I am attaching new debug logs, just in case.
[dbeaver-debug.log](https://github.com/dbeaver/dbeaver/files/11561361/dbeaver-debug.log)

**Comment by igor2x** [comment]
In DBeaver now I have deleted PostgreSQL connection and start from step 1 from my bug report and the problem is exactly the same as in originally reported problem.

Strange... I don't know what was wrong previously at my today's post at step 5. We can probably ignore it.

**Comment by igor2x** [comment]
Now I know what is error in step 5 above (when using password), if time on Linux where DBeaver is installed is way off the time of Kerberos server then error "_PREAUTH_FAILED with error code: KRB_AP_ERR_SKEW_" is returned. If I set time back to the time of Kerberos server, then Kerberos connection is successful.

Now back to my first post of official bug report. Still not working "kinit", like I have described above.

**Comment by LonwoLonwo** [comment]
Thanks for the investigation

According to this [discussion](https://stackoverflow.com/questions/64606520/org-postgresql-util-psqlexception-an-error-occurred-while-setting-up-the-gss-en), could you please change the value of this setting in the driver properties? Try different options.

![2023-05-25 12_15_24-Connect to a database](https://github.com/dbeaver/dbeaver/assets/45152336/40419b79-3752-4549-8d89-4a72e1f28532)

**Comment by igor2x** [comment]
### I. OFFICIAL DOCUMENTATION OPTIONS EXPLAINED
According to official JDBC driver documentation: https://jdbc.postgresql.org/documentation/use/
- **disable** is obvious and disables any attempt to connect using GSS encrypted mode
- **allow** will connect in plain text then if the server requests it will switch to encrypted mode
- **prefer** will attempt to connect in encrypted mode and fall back to plain text if it fails to acquire an encrypted connection
- **require** attempts to connect in encrypted mode and will fail to connect if that is not possible.

### II. POSTGRESQL CONFIGURATION
PostgreSQL is configured in _pg_hba.conf_ to only allow encrypted Kerberos connections from remote computers. Whole file looks like this:
```
# TYPE     DATABASE        USER            ADDRESS                 METHOD
hostgssenc all all all gss include_realm=0 krb_realm=EXAMPLE.COM

```

### III. TESTS WITH PASSWORD (as Step 5 in bug report):
Options **disable**  and **allow** returns:
![image](https://github.com/dbeaver/dbeaver/assets/10895030/008cdc42-3705-437e-a37f-04bd51906788)

Options  **prefer** and **require** returns:
![image](https://github.com/dbeaver/dbeaver/assets/10895030/7c6fbf89-3741-4007-a49c-16b0be934af3)

This works as expected, no issue here.

### IV. TESTS WITH KINIT (as Step 11 in bug report)

Options **disable**, **allow** and **prefer** returns:
![image](https://github.com/dbeaver/dbeaver/assets/10895030/008cdc42-3705-437e-a37f-04bd51906788)

Note: Above **prefer** is maybe not entirely expected. But I have checked the network packets with Wireshark to see what is going on. Prefer does starts with Kerberos then fails back to non-encrypted communication which also fails. Now DBeaver only reports last error (like picture above) and throws away error from Kerberos attempt. By the way, psql (default PostgreSQL client) in this case returns both errors (Kerberos and non-encrypted one). Here is maybe room for improvement, to actually display all of the connection errors, not just last one.

Option **require** returns (like reported in bug report):
![image](https://github.com/dbeaver/dbeaver/assets/10895030/e40a9267-253a-448f-9e91-9a19c1112b67)

**Comment by igor2x** [comment]
I did some additional tests.

1. In Linux shell I executed` kdestroy` command this deletes the TGT ticket. I tried to connect using psql client and as expected I get error message: _psql: error: connection to server at "server.example.com" (192.168.100.36), port 5432 failed: GSSAPI encryption required but was impossible (possibly no credential cache, no server support, or using a local socket)_

2. I tried with DBeaver and get the error: _An error occurred while setting up the GSS Encoded connection._

3. In Linux shell I created TGT ticket with `kinit user01` and tried psql connection, that finished successfully.

4. I tried with DBeaver and get the error as before: _An error occurred while setting up the GSS Encoded connection._

Looking into DBeaver debug logs and in both cases there are exactly the same messages (excluding the time of course). It looks to me that DBeaver fails somewhere entirely else or does not use the Linux kinit that I am using.

See debug logs:
Step 2: [dbeaver-debug_before.log](https://github.com/dbeaver/dbeaver/files/11564102/dbeaver-debug_before.log)
Step 4: [dbeaver-debug_after.log](https://github.com/dbeaver/dbeaver/files/11564101/dbeaver-debug_after.log)

**Comment by igor2x** [comment]
I did additional tests. From error messages and content of DBeaver logs I assumed DBeaver is unable to access TGT ticket.

1. I created new TGT ticket: `kinit -c /home/user01/krb5cc_user01` pointing TGT ticket to file instead of cache.

2. Tested in DBeaver:
![image](https://github.com/dbeaver/dbeaver/assets/10895030/f9c6eafe-a981-4fe3-8c25-5970fe984f4e)
and it fails with:
![image](https://github.com/dbeaver/dbeaver/assets/10895030/e40a9267-253a-448f-9e91-9a19c1112b67)

3. But typing in the same path as it is in step 1 in greyed text:
![image](https://github.com/dbeaver/dbeaver/assets/10895030/1fcc47b0-dfb0-49d4-9b20-27ed01737067)

4. And now connection is successful.

It looks like DBeaver is for some reason unable to read TGT ticket from cache and also default greyed text is not really useful (actually misleading).

---

To recap. It looks some issue with DBeaver unable to read TGT ticket from cache. That explains why DBeaver stops so early in the process of Kerberos protocol. Hope my analysis helps you a bit.

**Comment by igor2x** [comment]
I have investigated little bit more. When I create new TGT ticket using `kinit` command I get the following output:
![image](https://github.com/dbeaver/dbeaver/assets/10895030/2aa77219-f8d2-4e8a-8d19-e11d6e369f89)

I have checked the official MIT Kerberos documentation about [Credential cache](https://web.mit.edu/kerberos/krb5-latest/doc/basic/ccache_def.html) and the following is written: "**KEYRING is Linux-specific**, and uses the kernel keyring support to store credential data in unswappable kernel memory where only the current user should be able to access it."

Now the question is, does DBeaver supports Linux-specific KEYRING? If not, then at least currently the only option is not really secure, to use file for ticket cache.

**Comment by Matvey16** [comment]
@igor2x Hello, thank you for the report. Regarding the very first post whith the bug. Can you try connecting after changing the name from `user01` to `user01@EXAMPLE.COM`?

**Comment by igor2x** [comment]
All my above messages are for **DBeaver Lite** on **Ubuntu 22.04** accessing **MIT Kerberos** server at **kerberos.example.com**.

Today I have tried to do some additional tests and I see my trial version has expired, so I can't make additional tests. Our company is in the process of database tools evaluation that support Kerberos, so I can't buy it right now.

---

1. But... I have downloaded the latest **DBeaver Lite** v23.1.0 and installed in on **Windows 2022 Server** accessing **Active Directory** as KDC server that runs at different domain **kerberos.adexample.com**.

2. I have set **Driver properties** parameter to **gssEncMode** to **require**

3. Did the usual tests to enter Kerberos password.
![image](https://github.com/dbeaver/dbeaver/assets/10895030/0db77423-13cb-48c8-b2d4-8c656bb53fcf)

4. Cleared password and checked kinit
![image](https://github.com/dbeaver/dbeaver/assets/10895030/b60312a6-60a0-4623-b277-3674fcd67754)

5. From Windows CMD from %USERPROFILE% directory I have executed: `klist purge` to delete all Windows TGT tickets and repeated the process. The same error as in step 4. Repeated command: `klist` and not tickets are created.

6. I repeated step 3, and successful connection. Repeated `klist` command from user home directory and no ticket are created on Windows. This is strange. At this stage TGT ticket and TS tickers should exists. This is somehow a proof DBeaver does NOT uses operating system tickers. I assume DBeaver is using its own Kerberos libraries and not operating system Kerberos libraries.

7. I checked all of the DBeaver directories and in `%USERPROFILE%\AppData\Local\DBeaverLite\jre\bin` I have found kinit.exe and klist.exe commands. This looks like JRE delivers additional commands. If I execute `klist` command from above directory there are no tickers there in. If I create new TGT ticket from above directory: `kinit` type in password and I get: "New ticket is stored in cache file C:\Users\user01\krb5cc_user01". Back to DBeaver and clicked on "Browser" and selected created file.
![image](https://github.com/dbeaver/dbeaver/assets/10895030/5a3574d3-0357-493e-af96-4c69cfad7397)

and `Test Connection` button but DBeaver is still asking me for password. I don't really understand this. If TGT ticket is created DBeaver should not ask for password. It looks like DBeaver is not even using "kinit" from JRE. It looks to me DBeaver is establishing its own Kerberos connection to get TGT ticket.

---
I am willing to do some additional tests. But somehow it looks to me DBeaver does not uses operating system Kerberos implementation, but I assume it uses its own.

This same conclusion I came up in my last previous post on Ubuntu, that DBeaver does not uses operating system kinit implementation, but instead looks like some internal one (I assume calling some Java libraries).

---
Our company is looking for some database tool that is capable to use operating system level Kerberos. At the best I imagine in *Connection settings* dialog the only setting I should set in **Authentication** is to select "PostgreSQL Kerberos" all other info should be retrieved from operating system level Kerberos. **Kerberos user** is stored in TGT ticket (TGT ticket is created automatically on Windows when user logsin; on Ubuntu this is the same if single sign-on is implemented). Username shouldn't be needed because on PostgreSQL server pg_ident.conf file can be used to tranform Kerberos to PostgreSQL user. Like I don't see two separate users (Kerberos user and PostgreSQL user) when using psql command. Also **KDC Server** is stored in TGT ticket.

Maybe there should be additional option like:
- Use operating system level Kerberos: when checked all the input boxes should get greyed out and DBeaver is using operating system level Kerberos
- when unchecked all input boxes should appear just like now and user can set some additional settings if preferred.

Like I see it. Using Kerberos it should be way simpler for user to use database tool authentication, not like now way more difficult.

**Comment by LonwoLonwo** [comment]
Hello, again @igor2x 

Sorry for the late response.
Could you please check the latest DBeaver Early Acces version?
https://dbeaver.com/files/ea/
If the issue still occurs there, please create a ticket for our tech support https://dbeaver.com/tickets. You can add a link to this ticket there.

**Comment by igor2x** [comment]
We have made a lot of testing with PostgreSQL and Kerberos with many products and solutions (including DBeaver) and we see a lot of products and solutions have incorrect, buggy, partial support or not support at all for PostgreSQL/Kerberos.

If there is at least some kind of basic Kerberos support in products, there is pretty much huge problem when facing _Kerberos delegation_ - see for example [Youtube video](https://www.youtube.com/watch?v=UGWP4ewxcTA) what is Kerberos delegation. When delegation is in question, products are producing all kind of strange errors, which are very difficult to debug, because Kerberos is not really simple protocol. It looks like Kerberos is not seriously tested in any product (we have tested).

In our company decision was made that at current stage of Kerberos support in PostgreSQL in products and solutions we do **not** implement Kerberos as authentication mechanism, because we would be **forced** to use at least one more authentication method beside Kerberos, to support all our products.

Like I see it, currently the only truly working solution (from products we have tested) are the ones that are provided by PostgreSQL project itself, client and server site, all other is just not sufficiently supported for our needs.

P.S. Anyway, just out of curiosity I tried to do a quick test. I downloaded DBeaver early access version and after install I have got "license expired" error message. This is probably because few months ago I installed _DBeaver Light Edition_ **trial** version, and now I can't even test the early access version.
